### ISSUE TITLE
repair_history data not migrated along with tablet

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylladb#17507]
The repair history table records the last time repair was successful on a particular tablet, which in turn allows tombstone garbage collection to remove all older tombstones. However, when a tablet is migrated, this information is lost.

**Comment by denesb** [comment]
There are two possible solutions here:
1) Move repair_history to RAFT, make it a group0-distributed system table, there is a PR for this: https://github.com/scylladb/scylladb/pull/18211
2) Move repair_history content into the tablets table, which is already a group0-distributed table

I think (2) is a much better choice:
* Moving tables to group0 has a cost, we shouldn't just move everything to group0, if there are alternatives.
* system.tablets already exists and is managed by group0, storing per-repair history data in said table is a natural choice, and integrates nicely with incremental repair.
* If we keep using repair_history, we need to make it work for both tablets and vnodes. Best to leave repair_history as-is, to be vnode specific and use system.tablets for tablets.

**Comment by raphaelsc** [comment]
> There are two possible solutions here:
> 
> 1. Move repair_history to RAFT, make it a group0-distributed system table, there is a PR for this: [repair: Support repair_history migration after topology change #18211](https://github.com/scylladb/scylladb/pull/18211)
> 2. Move repair_history content into the tablets table, which is already a group0-distributed table
> 
> I think (2) is a much better choice:
> 
> * Moving tables to group0 has a cost, we shouldn't just move everything to group0, if there are alternatives.
> * system.tablets already exists and is managed by group0, storing per-repair history data in said table is a natural choice, and integrates nicely with incremental repair.
> * If we keep using repair_history, we need to make it work for both tablets and vnodes. Best to leave repair_history as-is, to be vnode specific and use system.tablets for tablets.

If we imposed a constraint of tablet repair having minimum granularity of one tablet range and only being considered successful once a given tablet range is successfully repaired, then we could simply have a last_repaired_timestamp in tablet metadata that would be available for all tablets wherever they go.

**Comment by denesb** [comment]
> If we imposed a constraint of tablet repair having minimum granularity of one tablet range and only being considered successful once a given tablet range is successfully repaired, then we could simply have a last_repaired_timestamp in tablet metadata that would be available for all tablets wherever they go.

Exactly, option 2 is as simple as adding a single new field to system.tablets.
I think that going forward, with automated and incremental tablet repair, we are going to enforce repairing one tablet at a time. And even if we don't enforce this, we simply don't update this timestamp when user goes against the recommendation and force-repairs half a tablet.

**Comment by asias** [comment]
> There are two possible solutions here:
> 
>     1. Move repair_history to RAFT, make it a group0-distributed system table, there is a PR for this: [repair: Support repair_history migration after topology change #18211](https://github.com/scylladb/scylladb/pull/18211)
> 
>     2. Move repair_history content into the tablets table, which is already a group0-distributed table
> 
> 
> I think (2) is a much better choice:
> 
>     * Moving tables to group0 has a cost, we shouldn't just move everything to group0, if there are alternatives.

The group0 just provides a service to distribute the data to nodes. Like rpc provides a service to send data.  It is not like we are managing the whole thing with group0. The amount of data is also tiny. 

>     * system.tablets already exists and is managed by group0, storing per-repair history data in said table is a natural choice, and integrates nicely with incremental repair.

system.tablets would only store the last repair time, while repair history stores the real history. The repair history is also useful.

> 
>     * If we keep using repair_history, we need to make it work for both tablets and vnodes. Best to leave repair_history as-is, to be vnode specific and use system.tablets for tablets.

Using repair_history for vnode and system.tablet for tablet separately means we need to keep both mechanisms forever until vnode support is dropped. Where to store repair time is only part of the story, we also need to update the in memory presentation of the repair history for compaction to query. Keeping the existing rpc based update + system.tablets based update means we need need two such mechanisms too. 

In addition, with option 2, vnode will still have the repair history is not migrated issue. With option 1, both tablet and vnode work, with a unified single mechanism to update repair history for tombstone gc. 

Using repair_history does not mean we could not have system.tablet.repair_time.  The PR https://github.com/scylladb/scylla-enterprise/pull/4239, is using system.tablet.repair_time for auto scheduling purpose. The system.tablet.repair_time is updated by the scheduler only.  If we update system.tablet.repair_time for manual repair, we need the very similar code like we do in  https://github.com/scylladb/scylladb/pull/18211 to system.tablet.repair_time though group0 after repair is  done. Now wee need to worry if the scheduling code or the low level repair code should update the repair_time and when to skip one of them. It is going to be a mess, not to mention we also need to do the extra check we could only update system.tablet.repair_time for user started repair if the whole range is updated.  

With option 2, repair ranges smaller than a tablet token range is not going to update repair history so tombstone gc repair is not going to work. We could argue this is fine but users might still repair smaller range in case a single range takes too long or for whatever reason. The extra limitation is gone if we use option 1.

**Comment by asias** [comment]
I added the release block label for the reason I explained here https://github.com/scylladb/scylladb/pull/18211#issuecomment-2516052117.

**Comment by mykaul** [comment]
> I added the release block label for the reason I explained here [#18211 (comment)](https://github.com/scylladb/scylladb/pull/18211#issuecomment-2516052117).

I don't disagree, but in this case, it needs to be worked on. @asias - are you on it?

**Comment by asias** [comment]
Yes.

**Comment by bhalevy** [comment]
@asias isn't this resolved now?

**Comment by asias** [comment]
> [@asias](https://github.com/asias) isn't this resolved now?

PR is published here: https://github.com/scylladb/scylladb/pull/21896
